string overhaul

StefanKarpinski opened the pull request:
Replacement PR for https://github.com/JuliaLang/julia/pull/24439. More description to come...

Edit see this discourse post for a description: https://discourse.julialang.org/t/changes-to-the-representation-of-char/7291/5

StefanKarpinski committed:
string overhaul: new Char representation, revamped core string API

StefanKarpinski committed:
LineEdit: use character syntax for wildcard character

StefanKarpinski committed:
strings/string: use checkbounds helper in a few places

StefanKarpinski committed:
repeat(Char, Integer): update & simplify for new Char rep

StefanKarpinski committed:
replace UnicodeError with StringIndexError

This type is appropriate to indicate that someone has indexed into
the non-leading code unit of a character in a string, regardless of
its encoding and the error is no longer used to indicate any other
Unicode-related errors, so the name was no longer fitting.

nalimilan commented on base/strings/basic.jl:
```diff
@@ -414,10 +455,7 @@ julia> chr2ind(str, 2)
-    s[i] # throws error if invalid
-    unsafe_ind2chr(s, i)
-end
+ind2chr(s::AbstractString, i::Integer) = length(s, 1, i)
```
Should we rename/remove `ind2chr` and `chr2ind`? There are discussions about renaming `sub2ind` and `ind2sub`, so the "2" pun could be completely removed from the API. Deprecations are trivial with the new methods.

StefanKarpinski replied:
Yes, that was my thought as well but I figured I'd leave it for another PR since this one is already so large.

nalimilan commented on base/strings/basic.jl:
```diff
@@ -585,16 +604,19 @@ julia> ucfirst("python")
-    return c==tc ? s : string(tc,s[nextind(s,1):end])
+    c′ = titlecase(c)
+    c == c′ ? convert(String, s) :
+    string(c′, SubString(s, nextind(s, 1)))
```
Fix indentation.

StefanKarpinski replied:
Do you prefer the "else" expression of a ternary to be indented or not – the above comment seems to indicate no but this comment seems to indicate yes...

nalimilan commented on base/strings/basic.jl:
```diff
@@ -585,16 +604,19 @@ julia> ucfirst("python")
-Return `string` with the first character converted to lowercase.
+Return `s` with the first character converted to lowercase.
+
+See also: `ucfirst`, `uppercase`, `lowercase`, `titlecase`
```
Should use `@ref` (here and elsewhere).

nalimilan commented on base/strings/string.jl:
```diff
@@ -1,7 +1,16 @@
+
 const ByteArray = Union{Vector{UInt8},Vector{Int8}}
 
+@inline between(b::T, lo::T, hi::T) where {T<:Integer} = (lo ≤ b) & (b ≤ hi)
```
Isn't this just `lo ≤ b ≤ hi`? If the problem is the difference between `&&` and `&`, I think the compiler avoids branching in simple cases like this anyway.

StefanKarpinski replied:
I was hoping so but the code gen was considerably different here. I would love to replace all uses of this with `lo ≤ b ≤ hi`.

StefanKarpinski replied:
Verified, these do generate different code:
```jl
julia> between1(x, lo, hi) = lo <= x <= hi
between1 (generic function with 1 method)

julia> between2(x, lo, hi) = (lo <= x) & (x <= hi)
between2 (generic function with 1 method)

julia> @code_native between1(2, 1, 10)
	.section	__TEXT,__text,regular,pure_instructions
; Function between1 {
; Location: REPL[78]:1
	cmpq	%rdi, %rsi
	jle	L8
	xorl	%eax, %eax
	retq
L8:
	cmpq	%rdx, %rdi
	setle	%al
	retq
	nop
;}

julia> @code_native between2(2, 1, 10)
	.section	__TEXT,__text,regular,pure_instructions
; Function between2 {
; Location: REPL[79]:1
	cmpq	%rdi, %rsi
	setle	%cl
	cmpq	%rdx, %rdi
	setle	%al
	andb	%cl, %al
	retq
	nop
;}
```
The performance difference seems to be negligible here, but I worry that in the context of loops and other complex code having branches will make it much harder for the compiler to optimize the surrounding code.

nalimilan commented on base/strings/string.jl:
```diff
@@ -208,134 +146,156 @@ byte_string_classify(s::String) =
+    @inbounds b = codeunit(s, i)
+    b & 0xc0 == 0x80 || @goto ret
+    u |= UInt32(b); i += 1
+@label ret
```
It's kind of weird to have no indentation for a label, given that it's only valid inside the parent function.

StefanKarpinski replied:
This is a pretty common style in C/C++ for goto labels, e.g. see http://en.cppreference.com/w/cpp/language/goto. The logic in my mind is that they are part of the control flow construct so having them outdented from the code seems fitting.

nalimilan commented on base/strings/string.jl:
```diff
@@ -208,134 +146,156 @@ byte_string_classify(s::String) =
+        while true
+            (i += 1) ≤ n || break
+            @inbounds b = codeunit(s, i) # lead byte
+        @label L
```
Same remark as above.

nalimilan commented on test/strings/basic.jl:
```diff
@@ -99,14 +99,14 @@ end
-    @test_throws BoundsError chr2ind("hello", 10)
-    @test_throws BoundsError ind2chr("hellø", 10)
-    @test_throws BoundsError chr2ind("hellø", 10)
+    @test ind2chr("hello", -1) == -1
```
Should we really support these? Why would -1 codeunit be equal to -1 character? That doesn't make a lot of sense.

nalimilan commented on test/strings/basic.jl:
```diff
@@ -612,17 +621,18 @@ end
-            @test nextind(strs[i], -1, 1) == 1
+            @test prevind(strs[i], 20) == 19
+            @test prevind(strs[i], 20, 1) == 19
+            @test prevind(strs[i], 20, 10) == 7
```
This kind of behavior sounds weird to me. So we assume that all imaginary characters take one code unit, and once we reach actual characters we use them? I can't imagine a situation where it would be useful, and it could hide bugs...

StefanKarpinski replied:
I originally wrote the assumption that they all take one code unit, but realized that for substrings it's very convenient to just pass-through to the backing string which means that the "imaginary" characters are whatever size the actual characters in the backing string are.

nalimilan replied:
So basically that means that the behavior of `prevind`/`nextind` is undefined when an out of bounds index is passed, which isn't very Julian IMHO. It would be more standard to throw an error when an out of bounds index is passed, and enable the undefined behavior via `@inbounds`, but as you said it's convenient in some cases, so...

StefanKarpinski replied:
It's not undefined – it satisfies lots of nice, predictable properties:

* `prevind(s, i) < i < nextind(s, i)`
* `prevind(s, i, n) ≤ i ≤ nextind(s, i, n)` for all `n ≥ 0`
* `prevind(s, nextind(s, i, n), n) == thisind(s, i)` for all `n > 0`
* `prevind(s, nextind(s, i, n), m) == nextind(s, i, n-m)` whenever `n > m`
* `nextind(s, prevind(s, i, n), m) == prevind(s, i, n-m)` whenever `n > m`

There is a very consistent philosophy to all of these changes: only throw an error when you _have_ to. For character iteration, the error only occurs when you try to ask about the code point of a malformed character – handling a character or a string that has malformed data is fine. For indexing, computing with out-of-bounds indices is fine, the only thing that's a problem is asking for a character or a string that is out of bounds. As I said before, we do this _all_ the time with array indices and it's not a problem this is just another form index arithmetic.

nalimilan replied:
I said it's undefined because the number of code units which are added depend on the parent string (if any), so for example it will be different for a plain `String` and for a `SubString`. So in practice if you pass an out-of-bounds index, call `nextind`/`prevind` a given number of times, and get an in-bounds index, it's totally unpredictable what character you get.

Array indices are quite different, since in that case the step is always 1.

StefanKarpinski replied:
Where did this out-of-bounds array index come from? If you start in bounds for a given string and then end back in bounds you always get the same answer. Yes, the out of bounds values in the middle may be different but why does that matter?

nalimilan replied:
Yes, normally you won't pass an out of bounds index. But if that happens because of a bug it would have been nice to be able to catch that by default.

StefanKarpinski replied:
Making out of bound code units always size one won’t help with that. We already discussed the bit about why allowing computing out of bounds values is useful. I guess what hasn’t been shown is that these functions need to accept out of bounds indices as well as produce them.

nalimilan commented on test/strings/types.jl:
```diff
@@ -46,10 +55,12 @@ for idx in [0, 1, 4]
 
-# second index beyond endof("∀∀")
-for idx in 5:8
+# index beyond endof("∀∀")
```
"second" was accurate.

StefanKarpinski replied:
I still don't get what it means.

nalimilan replied:
It refers to the second index (third argument) passed to `SubString`.

StefanKarpinski replied:
Oh, that was completely unclear to me since the comment doesn't mention `SubString` at all. I think I'd like to deprecate the three-arg constructor for `SubString` anyway and always use `SubString(s, i:j)` instead.

nalimilan reviewed (commented):
It's hard to review so many deep changes, but here are the few comments I can make.

StefanKarpinski commented on base/strings/basic.jl:
```diff
@@ -1,57 +1,188 @@
-sizeof(s::AbstractString) = error("type $(typeof(s)) has no canonical binary representation")
+isvalid(s::AbstractString, i::Integer) = typeof(i) === Int ?
+    throw(MethodError(isvalid, Tuple{typeof(s),Int})) :
+        isvalid(s, Int(i))
```
If there are two methods then any string type that defines `isvalid(s::StringType, i::Integer)` creates an ambiguity. We could not define any `isvalid(s::AbstractString, i::Integer)` method that converts to `Int` but that's annoying because then every type that implements the string interface has to implement the same convert to `Int` logic. This allows subtypes to just implement the `Int` method in the common case or handle `Integer` and either one is fine.

StefanKarpinski commented on base/iostream.jl:
```diff
@@ -452,14 +445,23 @@ function read(s::IOStream, nb::Integer; all::Bool=true)
+
 function peekchar(s::IOStream)
-    if ccall(:ios_peekutf8, Cint, (Ptr{Void}, Ptr{Char}), s, _chtmp) < 0
+    chref = Ref{UInt32}()
```
Not according to @Keno who suggested that I move the externally allocated box inside IIRC.

Keno replied:
welcome to the brave new world of julia 0.7

StefanKarpinski commented:
Regarding indexing past the ends of strings. Here's my reasoning for it. At the very least we need to support indices of `0` and `ncodeunits(s)+1` since those come up all the time in generic definitions and not supporting those leads to really awkward code to support edge cases. For example, the current definition of `ind2chr(s, i) = length(s, 1, i)` breaks down for `i == 0` unless we allow zero. Similarly, the definition of `chr2ind(s, i)` uses `prevind(s, 0, -n)` and `nextind(s, 0, n)`. Also consider the new simpler definitions of `first` as `s[1:min(end, nextind(s, 0, n))]` and `last` as `s[max(1, prevind(s, ncodeunits(s)+1, n)):end]`. I didn't end up needing it but there were moments where I was considering writing code like `nextind(s, -1, n+1)` for those. Instead, I relaxed the requirement that the number of characters passed to `nextind`/`prevind` to advance be strictly positive and allowed it to be zero, which ended up being simpler.

My general philosophy here (which I need to write down somewhere in the string docs) is that doing computations with indices outside of the bounds of strings is totally fine – it's only accessing a character past one of the ends of a string that's a problem. By analogy to indexing into an array, it's perfectly normal to compute intermediate values when doing index arithmetic that are outside of the range of the array. The problem is only if at the end you index into the array out of bounds. I ended up not needing to use indices outside of the `[0, ncodeunits(s)+1]` range into strings, but I find it a little hard to justify allowing those but being really strict about going outside of that range.

Here's a motivating example. Suppose you're writing code to find a particular word in some text with `k` characters of context around it. With the current "lax" behavior, you can find the word (say it goes from index `i` to `j` in `text`) and then compute the bounds of what you should show as `max(1, prevind(text, i, k))` and `min(endof(test), nextind(text, j, k))`. How do you write this code if `prevind` and `nextind` don't let you compute past the end of the string and throw an error instead? You would have to check if the beginning and end of `text` are each no more than 100 characters before and after the end of the word first. This is actually quite hard to do efficiently. You could do `length(text, 1, i)` but that's `O(length(text))` instead of `O(k)`. The only alternative I can think of is to do individual single-step `prevind` and `nextind` operations, checking at each one if you've hit the ends of the text or not. That's `O(k)` at least, but now you've lost any efficiency that the three-argument `prevind` and `nextind` might have for the string encoding, which, e.g. for the `String` type are considerable, since every `prevind` and `nextind` operation needs to figure out if the index it's at is the beginning of a character or not, which is now slightly non-trivial.

What about having `prevind` and `nextind` allow you to advance beyond the ends of a string but just saturating at `0` and `ncodeunits(s)+1`? That has all the problems that saturating arithmetic for normal integers does. In the case of UTF-32 (or any fixed-width encoding) where these operations are just normal integer arithmetic, having an API where index operations saturate or throw errors means that you can't have the benefits of simple indexing – instead you're forced to implement `nextind` in a complicated way that prevents optimizations rather than `nextind(s, i, n) = i + n` and `prevind(s, i, n) = i - n`. Even in encodings where `nextind` and `prevind` can't be that simple, saturation ruins all the nice universal properties you might want for these functions, such as that `prevind(s, i) < i < nextind(s, i)` or that `prevind(s, nextind(s, i, n), n) == thisind(s, i)` and so on. Instead, you have to throw in caveats about the ends of strings and whatnot.

nalimilan commented:
OK, I see the advantages. But I still still find it disturbing when `nextind` or `prevind` start with an out-of-bounds character, iterate over a few imaginary one-codeunit characters, and enter the actual string at some point. But well...

StefanKarpinski commented:
Can you think of an efficient way to handle the word search with context example any other way?

StefanKarpinski commented on base/strings/basic.jl:
```diff
@@ -1,57 +1,188 @@
+a character in `s` or not. If `isvalid(s, i)` is true then `s[i]` will return
+the character whose encoding starts at that index, if it's false, then `s[i]`
+will raise an invalid index error. Behavior of `next(s, i)` is similar except
+that the character is returned along with the index of the following character.
```
I think there's a fair amount of generic string code in Base (and presumably elsewhere) that assumes that the string iteration state is the same as the string index. If it's not safe to assume that then we need to make sure that people don't use the iterator state in `s[i]`.

iamed2 replied:
I would very much like the ability to override string iteration for MyStringType (which I believe is currently possible?), and have the docs reflect the default behaviour of `next` if you don't also override `start` and `done`.

StefanKarpinski replied:
@iamed2, I'm not quite sure what you're asking for – do you want to allow the iteration state not to be the string index? If so, I'm totally fine with that and it's mostly a documentation issue; for the built-in string types they happen to coincide. Might be worth changing the iteration state of `SubString` to match its parent's iteration state just to shake out the places where we currently rely on these being the same though.

vtjnash replied:
This'll need to be updated in that case to provide whatever is most efficient:
https://github.com/JuliaLang/julia/blob/224f1e9f8666545a5e3e0a7be721231df88e0788/base/io.jl#L649-L650
Pretty easy just to add the offset as needed, if that's what we want. At the time, I had seen that in a number of other places we define the bounds of the strings indices to be `1:n`, so just went with it:
https://github.com/JuliaLang/julia/blob/224f1e9f8666545a5e3e0a7be721231df88e0788/base/strings/basic.jl#L24-L25

StefanKarpinski replied:
Yes, those are sane fallback definitions. Most strings work that way. The question is just that _if_ someone overrides those, do we support it working. Note that this is not actually pressing for 1.0 since if we don't allow it now, we can always allow it in the future. However, changing the iteration state of substrings might break people's code, so we would want to do that now.

iamed2 replied:
> @iamed2, I'm not quite sure what you're asking for – do you want to allow the iteration state not to be the string index? If so, I'm totally fine with that and it's mostly a documentation issue; for the built-in string types they happen to coincide. Might be worth changing the iteration state of SubString to match its parent's iteration state just to shake out the places where we currently rely on these being the same though.

Yes, exactly :)

StefanKarpinski commented:
I was tempted to just make `StringIndexError` a `KeyError` but the name sounds a bit funny. If `KeyError` were called `IndexError` it would work a bit better. We have a lot of almost-overlapping error types for bad `i` in `obj[i]`, I have to wonder if we can't unify them a bit.

StefanKarpinski committed:
See also: use `[name](@ref)` cross reference links.

StefanKarpinski committed:
use stevengj's suggested `hash(Char)` definition

StefanKarpinski committed:
address Milan's review comments

nalimilan commented:
> Can you think of an efficient way to handle the word-search-with-context example any other way?

Well, it would work if `prevind`/`nextind` stopped when reaching the bounds of the string. But as you noted that behavior has drawbacks too, but in practical terms I'm not sure whether they are really problematic or not.

> I was tempted to just make StringIndexError a KeyError but the name sounds a bit funny. If KeyError were called IndexError it would work a bit better. We have a lot of almost-overlapping error types for bad i in obj[i], I have to wonder if we can't unify them a bit.

Given how special string indices are, it makes sense to have a dedicated exception for them.

StefanKarpinski commented:
> But as you noted that behavior has drawbacks too, but in practical terms I'm not sure whether they are really problematic or not.

I suspect I can come up with a slightly more complex use case where coming back in-bounds is required for the above word-with-context example. But there's a reason we don't do saturating integer arithmetic and this is just another form of integer arithmetic.

StefanKarpinski commented:
I'm getting
```jl
 !! No doc found for reference '[`ncodeunits`](@ref)'.
```
Anyone know why that would be happening? There is a docstring for `ncodeunits` so I'm confused about what the problem is.

nalimilan commented:
I think you need to add the function to the stdlib part of the manual for Documenter to find it.

StefanKarpinski committed:
add ncodeunits to stdlib docs

StefanKarpinski committed:
strings: some formatting tweaks

StefanKarpinski commented:
@nalimilan, I've thought about this a bit and while there are examples (like the word-context one) that show that it's useful to _produce_ out-of-bounds indices from string index arithmetic functions, I can't think of a concrete use case for _accepting_ them – except for the only-just-out-of-bounds ones, i.e. `0` and `ncodeunits(s)+1`. If you have some code that produces a very out-of-bounds index and then passes that to another string index arithmetic function, then why not just combine those operations? The identities I showed above demonstrate that you can do since

* `prevind(s, nextind(s, i, n), m) == nextind(s, i, n - m)` whenever `n > m`
* `nextind(s, prevind(s, i, n), m) == prevind(s, i, n - m)` whenever `n > m`

So why not replace the RHS with the LHS which does less work and may avoid going out of bounds where the LHS went out of bounds temporarily? So I guess the conservative thing to do where would be to make these functions raise errors when passed starting indices outside of the range `0:ncodeunits(s)+1` even though they are allowed to _produce_ out of bounds values. It bothers me that some out-of-bounds indices are allowed while others aren't. What principled reason do we have for that distinction other than that it's practical? In my view producing very out of bounds indices is safer than saturating at `0` and `ncodeunits(s)+1` since saturating might very well hide a computation where you end up back in bounds but should not have – it's better to produce very out of bounds values which are then an error to use.

StefanKarpinski commented:
Related issue: `isvalid`. Previously, `isvalid` simply returned `false` for out-of-bounds indices. In this PR I've made a clearer distinction between out-of-bounds indices – i.e. those not satisfying `1 ≤ i ≤ ncodeunits(s)` – and invalid indices that are not the start of a character. The new `isvalid` throws an error if given an out-of-bounds index. I'm not sure what the better behavior is.

StefanKarpinski committed:
delete bswap(::Char)

StefanKarpinski commented:
I've pushed a new commit deleting the `bswap(::Char)` method. I had added a deprecation but when I want to test the deprecation, I realized that there are basically no character values (`'\0'` is the only exception) for which you can bswap the 32-bit code point and get back a valid code point, so this function is essentially completely useless and cannot really be made to work in the new scheme since there's no way to take a completely invalid code point and turn it into a `Char` value.

nalimilan commented:
It sounds reasonable to start with the most restrictive behavior (throwing an error), since if it turns out to be problematic in some cases we will be able to change it, which wouldn't be possible in the other direction.

StefanKarpinski commented:
Yeah, I suppose so, but it just feels a bit weird and undisciplined to allow some out-of-bounds indices – `0` and `ncodeunits(s)+1` – but not others – `-1` or `ncodeunits(s)+2`. And we have very real use cases for the former so we don't want to disallow those.

StefanKarpinski committed:
deprecate chr2ind and ind2chr

StefanKarpinski commented:
I have now added the deprecation of `ind2chr` and `chr2ind` in a separate commit (for easier review). It was fairly straightforward and improves performance in a few places. In the process I also encountered the https://github.com/JuliaLang/julia/issues/25016 `lpad` and `rpad` issue. I've updated the code to those functions but left the inconsistent behavior in-place for now.

nalimilan commented on doc/src/manual/strings.md:
```diff
@@ -565,14 +565,15 @@ Some other useful functions include:
 
   * [`endof(str)`](@ref) gives the maximal (byte) index that can be used to index into `str`.
   * [`length(str)`](@ref) the number of characters in `str`.
+  * [`length(str, i, j)`](@ref) the number of valid character indices in `str` from `i` to `j`.
```
Actually I wonder whether we should add this method, as it looks a bit weird. Would it be enough to recommend `length(SubString(str, i:j))`?

vtjnash replied:
It likely can be free if it gets inlined. But seem to recall the SubString constructor does quite a bit of work to munge the indices into something valid? Anyways, doesn't seem like this API decision should be too much influenced by temporary performance concerns (unless it's only for internal usage, but them maybe call it `Base.substr_length`).

GregPlowman commented:
> I'm getting
> 
>  !! No doc found for reference '[`ncodeunits`](@ref)'.
> Anyone know why that would be happening? There is a docstring for ncodeunits so I'm confused about what the problem is. I seem to be using the @ref thing wrong somehow although as far as I can tell I'm using the same way as all the other docstrings do.

Not sure if its related, but I had a similar problem: https://github.com/JuliaLang/julia/pull/23703#issuecomment-332702298

Removing the backticks fixed the issue, but this seemed less than ideal because I thought "auto-referencing" should work with or without backticks.

StefanKarpinski committed:
move string search functions into strings/search.jl

StefanKarpinski committed:
optimize the length(::String) method better

StefanKarpinski committed:
make length(string, i, j) ≥ 0 for all i, j

StefanKarpinski committed:
moar Unicode operators

StefanKarpinski committed:
various String performance tweaks

- {next,getindex}_continued don't need to re-check bounds
- short-circuiting is slightly faster in length

StefanKarpinski commented:
@nalimilan: is it ok with you if I merge my PR as is to avoid it getting lots of conflicts and then make the changes to make indexing more conservative in a separate PR? My general plan is to try this and see how it goes:
- make `isvalid` return false for out-of-bounds indices instead of an error
- make `thisind`, `nextind`, `prevind`, `length` error on most/all out-of-bounds indices
  - allow `0` for `nextind` and the `lo` argument to `length`
  - allow `ncodeunits(s)+1` for `prevind` and the `hi` argument of `length`
  - otherwise only allow in-bounds indices

StefanKarpinski committed:
bounds checks on string length(s, i, j)

StefanKarpinski committed:
change some code point conversions to char comparisons

StefanKarpinski committed:
add various `@propagate_inbounds` annotations

also: avoid second bounds check in `get` with default

StefanKarpinski committed:
isvalid: return false out of bounds instead of throwing

also: `next` can assume that incoming indices are valid

tpapp commented on base/strings/basic.jl:
```diff
@@ -1,60 +1,210 @@
+ncodeunits(s::AbstractString)
+
+"""
+    codeunit(s::AbstractString) -> Type{<:Union{UInt8, UInt16, UInt32}}
```
(from a [discussion](https://discourse.julialang.org/t/changes-to-the-representation-of-char/7291/9?u=tamas_papp) on Discourse)

I think that the function that retrieves the type should perhaps have a different name than the function that returns a codeunit. eg `codeunit_type`. Methods of the same function usually share behavior on some abstract level, this is stretching that. I realize that the name is long, but it is only used in a few places (mostly to define the method for various types). I expect that in code outside `Base` it will be only used to help type inference so that this will not be cumbersome.

nalimilan commented:
Sorry, I hadn't seen your question. Yes, that's fine with me. (Actually, what worries me more are the conflicts with the search/find PRs, but such is life...)

StefanKarpinski commented:
I've been continuing to work here, so I guess I'll just have a gnarly rebase after you merge that.

StefanKarpinski committed:
bounds check thisind, nextind and prevind as well

StefanKarpinski commented:
Ok, I think this is good to go as soon as CI passes. There's still the issue of potentially renaming `codeunit(s::AbstractString)` to `codeunittype` but I'm not 100% convinced and that's an easy change to make separately (and almost no one will actually have used that function).

StefanKarpinski commented:
The index bounds rules I ended up going with (which still need to be documented):

* `thisind(s, i)`: `0 ≤ i ≤ n+1`
* `nextind(s, i)`: `0 ≤ i < n+1`
* `prevind(s, i)`: `0 < i ≤ n+1`
* `length(s, i, j)`: `0 < i ≤ n+1`, `0 ≤ j < n+1`

The logic is that if you're already out of bounds in some direction, you shouldn't ask to go *more* out of bounds in that direction. For `thisind` it's ok to pass a just-out-of-bounds index and get back the same index. This was pretty necessary for the ways that `thisind` gets used. For `length`, the logic is that the start index can be out of bounds past the end of the string while the end index can be out of bounds past the start of the string, but not vice versa – otherwise you're asking to count non-existent characters. One oddity of these rules for `length` is that `length("", 1, 1)` is what `ind2chr("", 1)` is deprecated to and it is now a bounds error because of the `j = 1` index. I didn't encounter any practical issues with this, but it's possible that the bounds for length should be slightly relaxed.

StefanKarpinski commented:
Oy, those conflicts...

StefanKarpinski commented:
Is the memory error on FreeBSD expected these days? cc @iblis17

StefanKarpinski commented:
Ok, I've already tested the heck out of this change, the BSD failure seems to be an unrelated memory problem, it passes on Windows and Linux and I can't reproduce the macOS spawn failure locally, so I'm assuming that's unrelated too.

StefanKarpinski commented:
I did not mean to squash that, but I think it's fine. The intermediate commits are not all that useful.

iblis17 commented:
@StefanKarpinski about the ENOMEM: https://github.com/JuliaLang/julia/pull/23273#issuecomment-348415871

StefanKarpinski commented:
Actually this wasn't squashed, GitHub just changed how the display merge commits.
